File: InteractHub.API/Services/Implementations/PostService.cs
using System.Text.RegularExpressions;
using InteractHub.API.DTOs.Posts;
using InteractHub.API.Entities;
using InteractHub.API.Repositories.Interfaces;
using InteractHub.API.Services.Interfaces;

namespace InteractHub.API.Services.Implementations;

public class PostService : IPostService
{
    private readonly IPostRepository _postRepo;
    private readonly IMediaService _mediaService;
    private readonly IHashtagService _hashtagService;
    private readonly IPostHashtagRepository _postHashtagRepo;

    public PostService(
        IPostRepository postRepo,
        IMediaService mediaService,
        IHashtagService hashtagService,
        IPostHashtagRepository postHashtagRepo)
    {
        _postRepo = postRepo;
        _mediaService = mediaService;
        _hashtagService = hashtagService;
        _postHashtagRepo = postHashtagRepo;
    }

    // ================= CREATE =================
    public async Task<PostResponseDto> CreatePostAsync(string userId, PostCreateDto dto)
    {
        var post = new Post
        {
            UserId = userId,
            Title = dto.Title,
            Content = dto.Content,
            Status = dto.Status ?? 1,
            CreatedAt = DateTime.UtcNow,
            PostMedias = new List<PostMedia>()
        };

        // Upload media
        if (dto.Files != null && dto.Files.Any())
        {
            foreach (var file in dto.Files)
            {
                var url = await _mediaService.SaveFileAsync(file, "posts");

                if (!string.IsNullOrEmpty(url))
                {
                    post.PostMedias.Add(new PostMedia
                    {
                        Url = url,
                        MediaType = file.ContentType.Contains("video") ? 1 : 0
                    });
                }
            }
        }

        await _postRepo.AddAsync(post);
        await _postRepo.SaveChangesAsync();

        // 🔥 HANDLE HASHTAG
        await HandleHashtagsAsync(post.Id, post.Content);

        var result = await _postRepo.GetPostDetailsByIdAsync(post.Id);
        return MapToDto(result ?? post);
    }

    // ================= UPDATE =================
    public async Task<PostResponseDto> UpdatePostAsync(int postId, string userId, PostUpdateDto dto)
    {
        var post = await _postRepo.GetPostDetailsByIdAsync(postId);
        if (post == null) throw new Exception("Bài viết không tồn tại.");
        if (post.UserId != userId) throw new Exception("Không có quyền.");

        if (dto.Content != null) post.Content = dto.Content;
        if (dto.Status != null) post.Status = dto.Status.Value;

        // XÓA MEDIA
        if (dto.DeleteMediaUrls != null && dto.DeleteMediaUrls.Any())
        {
            var deleteList = post.PostMedias
                .Where(m => dto.DeleteMediaUrls.Contains(m.Url))
                .ToList();

            foreach (var m in deleteList)
            {
                _mediaService.DeleteFile(m.Url);
                post.PostMedias.Remove(m);
            }
        }

        // THÊM MEDIA
        if (dto.NewFiles != null && dto.NewFiles.Any())
        {
            foreach (var file in dto.NewFiles)
            {
                var url = await _mediaService.SaveFileAsync(file, "posts");

                if (!string.IsNullOrEmpty(url))
                {
                    post.PostMedias.Add(new PostMedia
                    {
                        Url = url,
                        MediaType = file.ContentType.Contains("video") ? 1 : 0
                    });
                }
            }
        }

        // 🔥 UPDATE HASHTAG
        if (dto.Content != null)
        {
            await _postHashtagRepo.DeleteByPostIdAsync(post.Id);
            await HandleHashtagsAsync(post.Id, post.Content);
        }

        await _postRepo.SaveChangesAsync();

        var updated = await _postRepo.GetPostDetailsByIdAsync(post.Id);
        return MapToDto(updated ?? post);
    }

    // ================= DELETE =================
    public async Task DeletePostAsync(int postId, string userId)
    {
        var post = await _postRepo.GetPostDetailsByIdAsync(postId);
        if (post == null) throw new Exception("Không tìm thấy.");
        if (post.UserId != userId) throw new Exception("Không có quyền.");

        if (post.PostMedias != null)
        {
            foreach (var m in post.PostMedias)
            {
                _mediaService.DeleteFile(m.Url);
            }
        }

        await _postHashtagRepo.DeleteByPostIdAsync(post.Id);

        _postRepo.Delete(post);
        await _postRepo.SaveChangesAsync();
    }

    // ================= GET =================
    public async Task<IEnumerable<PostResponseDto>> GetTimelineAsync()
    {
        var posts = await _postRepo.GetPostsWithDetailsAsync();
        return posts.Select(MapToDto);
    }

    public async Task<IEnumerable<PostResponseDto>> GetPostsByUserIdAsync(string userId)
    {
        var posts = await _postRepo.GetPostsByUserIdAsync(userId);
        return posts.Select(MapToDto);
    }

    public async Task<PostResponseDto> GetPostByIdAsync(int postId)
    {
        var post = await _postRepo.GetPostDetailsByIdAsync(postId);
        if (post == null) throw new Exception("Không tồn tại.");
        return MapToDto(post);
    }

    // ================= HASHTAG CORE =================
    private async Task HandleHashtagsAsync(int postId, string? content)
{
    if (string.IsNullOrWhiteSpace(content)) return;

    // 👉 Dùng service đã có (đã xử lý: extract + create nếu chưa tồn tại)
    var hashtags = await _hashtagService.ExtractHashtagsAsync(content);

    if (hashtags == null || !hashtags.Any()) return;

    // 👉 Map sang bảng Post_Hashtag
    var mappings = hashtags.Select(tag => new Post_Hashtag
    {
        PostId = postId,
        HashtagId = tag.Id
    }).ToList();

    await _postHashtagRepo.AddRangeAsync(mappings);
    await _postHashtagRepo.SaveChangesAsync();
}

    // ================= DTO =================
    private static PostResponseDto MapToDto(Post p) => new()
    {
        Id = p.Id,
        Title = p.Title,
        Content = p.Content,
        UserId = p.UserId,
        Status = p.Status,
        AuthorName = p.User?.FullName ?? p.User?.UserName ?? "User",
        AuthorAvatar = p.User?.ProfilePicture ?? "",
        CreatedAt = p.CreatedAt,
        MediaUrls = p.PostMedias?.Select(m => m.Url).ToList() ?? new List<string>()
    };
}

File: InteractHub.Client/src/services/postService.ts
import axiosInstance from "./axiosInstance";

// ── 1. Interface BE (PascalCase - Khớp với C# PostResponseDto) ──
interface PostResponseDto {
  Id: number;
  Title?: string;
  Content?: string;
  UserId?: string;
  Status?: number;
  AuthorName?: string;
  AuthorAvatar?: string;
  CreatedAt?: string;
  MediaUrls: string[];
}
// ── 3. Interface FE (camelCase) ──────────────────────────────────
export interface PostItem {
  id: number;
  title?: string;
  content?: string;
  userId?: string;
  status?: number;
  authorName?: string;
  authorAvatar?: string;
  createdAt?: string;
  mediaUrls: string[];
}

export interface PostReportItem {
  id: number;
  postId: number;
  userId: string;
  userName: string;
  reason: string;
  status: number;
  createdAt: string;
}

export interface PostReportRequest {
  postId: number;
  reason: string;
}

// ── 4. Map BE → FE ──────────────────────────────────────────────
const mapPost = (p: PostResponseDto): PostItem => ({
  id: p.Id,
  title: p.Title,
  content: p.Content,
  userId: p.UserId,
  status: p.Status,
  authorName: p.AuthorName,
  authorAvatar: p.AuthorAvatar,
  createdAt: p.CreatedAt,
  mediaUrls: p.MediaUrls || [],
});



// ── 5. Export Service ───────────────────────────────────────────
export const postService = {
  createPost: (formData: FormData) =>
    axiosInstance
      .post<{ Success: boolean; Message: string; Data: PostResponseDto }>("/api/post/create", formData, {
        headers: { "Content-Type": "multipart/form-data" },
      })
      .then((res) => ({ ...res, data: mapPost(res.data.Data) })),

  getAllPosts: () =>
    axiosInstance
      .get<{ Success: boolean; Message: string; Data: PostResponseDto[] }>("/api/post/all")
      .then((res) => ({ ...res, data: res.data.Data.map(mapPost) })),


  getPostsByUserId: (userId: string) =>
    axiosInstance
      .get<{ Success: boolean; Message: string; Data: PostResponseDto[] }>(`/api/post/user/${userId}`)
      .then((res) => ({ ...res, data: res.data.Data.map(mapPost) })),

  getPostById: (postId: number) =>
    axiosInstance
      .get<{ Success: boolean; Message: string; Data: PostResponseDto }>(
        `/api/post/${postId}`
      )
      .then((res) => ({ ...res, data: mapPost(res.data.Data) })),

  deletePost: (postId: number) =>
    axiosInstance
      .delete<{ Success: boolean; Message: string }>(
        `/api/post/delete/${postId}`
      )
      .then((res) => ({
        success: res.data.Success,
        message: res.data.Message,
      })),

  updatePost: (postId: number, formData: FormData) =>
    axiosInstance
      .put<{ Success: boolean; Message: string; Data: PostResponseDto }>(`/api/post/update/${postId}`, formData, {
        headers: { "Content-Type": "multipart/form-data" },
      })
      .then((res) => ({ ...res, data: mapPost(res.data.Data) })),
  reportPost: (request: PostReportRequest) =>
    axiosInstance
      .post<{ Success: boolean; Message: string }>("/api/post-reports", request),
};